File: MosaicProcessor.cpp
#include "MosaicProcessor.h"
#include <iostream>
#include <algorithm>
#include <numeric>
#include <cmath>

//структура FeatureUtils
//вспомогательные функции вычисления признаков

//вычисляет стандартное отклонение (контрастность) изображения по каналам
cv::Scalar FeatureUtils::computeStdDev(const cv::Mat& image) {
    cv::Mat mean, stddev;
    
    cv::meanStdDev(image, mean, stddev);//вычисляет среднее значение и стандартное отклонение по каналам

    if (image.channels() == 3) {//для цветного BGR изображения возвращаем отклонения по всем трем каналам
        //возвращает скаляр со значениями стандартного отклонения для каждого канала
        return cv::Scalar(stddev.at<double>(0, 0), 
            stddev.at<double>(1, 0),
            stddev.at<double>(2, 0));
    }
    //для одноканального изображения возвращаем одно значение
    return cv::Scalar(stddev.at<double>(0, 0));
}

//вычисляет гистограмму направлений градиентов(HOG - подобный признак)
//гистограмма показывает распределение направлений краев в изображении
cv::Mat FeatureUtils::computeGradientHist(const cv::Mat& image) {
    //проверка входного изображения
    if (image.empty() || image.rows < 3 || image.cols < 3) {
        return cv::Mat::zeros(36, 1, CV_32F);
    }
    cv::Mat gray, grad_x, grad_y, magnitude, angle;
    //преобразование в оттенки серого для цветных изображений
    if (image.channels() == 3) {
        cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
    }
    else {
        gray = image;
    }
    gray.convertTo(gray, CV_32F);
    //вычисление градиентов по осям X и Y с помощью оператора Собеля
    cv::Sobel(gray, grad_x, CV_32F, 1, 0, 3);
    cv::Sobel(gray, grad_y, CV_32F, 0, 1, 3);
    //преобразование декартовых координат в полярные (величина и угол)
    cv::cartToPolar(grad_x, grad_y, magnitude, angle, true);
    //создание гистограммы с 36 бинами (по 10 градусов на бин)
    const int histSize = 36;
    cv::Mat hist = cv::Mat::zeros(histSize, 1, CV_32F);
    float angle_step = 360.0f / histSize;
    //заполнение гистограммы величинами градиентов
    for (int y = 0; y < angle.rows; ++y) {
        for (int x = 0; x < angle.cols; ++x) {
            float angle_val = angle.at<float>(y, x);
            float mag_val = magnitude.at<float>(y, x);
            int bin = static_cast<int>(angle_val / angle_step);
            bin = std::max(0, std::min(bin, histSize - 1));
            hist.at<float>(bin) += mag_val;
        }
    }
    //нормализация гистограммы
    cv::normalize(hist, hist, 1.0, 0.0, cv::NORM_L1);
    hist.convertTo(hist, CV_32F);
    return hist;
}

//метод для вычисления LBP-значения для пикселя
//LBP (Local Binary Pattern) - метод описания текстуры изображения
uchar FeatureUtils::getLBPValue(const cv::Mat& gray, int r, int c) {
    uchar center = gray.at<uchar>(r, c);
    uchar value = 0;
    //сравнение центрального пикселя с 8 соседями и формирование бинарного кода
    value |= (gray.at<uchar>(r - 1, c - 1) > center) << 7;
    value |= (gray.at<uchar>(r - 1, c) > center) << 6;
    value |= (gray.at<uchar>(r - 1, c + 1) > center) << 5;
    value |= (gray.at<uchar>(r, c + 1) > center) << 4;
    value |= (gray.at<uchar>(r + 1, c + 1) > center) << 3;
    value |= (gray.at<uchar>(r + 1, c) > center) << 2;
    value |= (gray.at<uchar>(r + 1, c - 1) > center) << 1;
    value |= (gray.at<uchar>(r, c - 1) > center) << 0;
    return value;
}

//вычисляет LBP-признаки для описания текстуры изображения
//возвращает гистограмму LBP-кодов
cv::Mat FeatureUtils::computeLBPFeatures(const cv::Mat& image) {
    //проверка входного изображения
    if (image.empty() || image.rows < 3 || image.cols < 3) {
        return cv::Mat::zeros(256, 1, CV_32F);
    }
    //преобразование в оттенки серого для цветных изображений
    cv::Mat gray;
    if (image.channels() == 3) {
        cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
    }
    else {
        gray = image;
    }
    //создание LBP-карты
    cv::Mat lbp = cv::Mat::zeros(gray.rows - 2, gray.cols - 2, CV_8UC1);
    for (int i = 1; i < gray.rows - 1; i++) {
        for (int j = 1; j < gray.cols - 1; j++) {
            lbp.at<uchar>(i - 1, j - 1) = getLBPValue(gray, i, j);
        }
    }
    //вычисление гистограммы LBP-кодов
    const int histSize = 256;
    float range[] = { 0, 256 };
    const float* ranges[] = { range };
    cv::Mat hist;
    cv::calcHist(&lbp, 1, 0, cv::Mat(), hist, 1, &histSize, ranges, true, false);
    //нормализация гистограммы
    cv::normalize(hist, hist, 1.0, 0.0, cv::NORM_L1);
    hist.convertTo(hist, CV_32F);
    return hist;
}

//класс ColorMetric
//вычисляет параметры клетки
void ColorMetric::computeCellFeatures(Tile& cell, const cv::Mat& cellImage) {
    cell.color = cv::mean(cellImage);
}
//вычисляет параметры тайтла
void ColorMetric::computeTileFeatures(Tile& tile, const cv::Mat& tileImage) {
    tile.color = cv::mean(tileImage);
}
//вычисляет расстояние между параметрами клетки и тайла на основе цвета
double ColorMetric::distance(const Tile& cell, const Tile& tile) const {
    return cv::norm(cell.color, tile.color);
}
//геттер для получения имени метрики
std::string ColorMetric::getName() const {
    return "color";
}

//класс ColorContrastMetric
//вычисляет параметры клетки
void ColorContrastMetric::computeCellFeatures(Tile& cell, const cv::Mat& cellImage) {
    cell.color = cv::mean(cellImage);
    cell.stddev = FeatureUtils::computeStdDev(cellImage);
}
//вычисляет параметры тайтла
void ColorContrastMetric::computeTileFeatures(Tile& tile, const cv::Mat& tileImage) {
    tile.color = cv::mean(tileImage);
    tile.stddev = FeatureUtils::computeStdDev(tileImage);
}
//вычисляет расстояние между параметрами клетки и тайла на основе цвета и контрастности
double ColorContrastMetric::distance(const Tile& cell, const Tile& tile) const {
    double colorDist = cv::norm(cell.color, tile.color);
    double stddevDist = cv::norm(cell.stddev, tile.stddev);
    return colorDist + 2.0 * stddevDist;
}
//геттер для получения имени метрики
std::string ColorContrastMetric::getName() const {
    return "color_contrast";
}

//класс GradientMetric
//вычисляет параметры клетки
void GradientMetric::computeCellFeatures(Tile& cell, const cv::Mat& cellImage) {
    cell.gradientHist = FeatureUtils::computeGradientHist(cellImage);
}
//вычисляет параметры тайтла
void GradientMetric::computeTileFeatures(Tile& tile, const cv::Mat& tileImage) {
    tile.gradientHist = FeatureUtils::computeGradientHist(tileImage);
}
//вычисляет расстояние между параметрами клетки и тайла на основе гистограмм градиентов
double GradientMetric::distance(const Tile& cell, const Tile& tile) const {
    //проверка гистограмм (не пустые)
    if (cell.gradientHist.empty() || tile.gradientHist.empty()) {
        return std::numeric_limits<double>::max();
    }
    //вычисление расстояния Бхаттачарии между гистограммами
    double histDist = cv::compareHist(cell.gradientHist, tile.gradientHist, cv::HISTCMP_BHATTACHARYYA);

    // Масштабируем до диапазона, сравнимого с цветовыми метриками
    return histDist * 1000.0;
}
//геттер для получения имени метрики
std::string GradientMetric::getName() const {
    return "gradient";
}

//класс TextureMetric
//вычисляет параметры клетки
void TextureMetric::computeCellFeatures(Tile& cell, const cv::Mat& cellImage) {
    cell.textureFeatures = FeatureUtils::computeLBPFeatures(cellImage);
}
//вычисляет параметры тайтла
void TextureMetric::computeTileFeatures(Tile& tile, const cv::Mat& tileImage) {
    tile.textureFeatures = FeatureUtils::computeLBPFeatures(tileImage);
}
//вычисляет расстояние между параметрами клетки и тайла на основе текстурных признаков
double TextureMetric::distance(const Tile& cell, const Tile& tile) const {
    const cv::Mat& hist1 = cell.textureFeatures;
    const cv::Mat& hist2 = tile.textureFeatures;
    //проверка гистограмм
    if (hist1.empty() || hist2.empty() || hist1.size() != hist2.size()) {
        return std::numeric_limits<double>::max();
    }
    //вычисление расстояния Бхаттачарии между гистограммами LBP
    double histDist = cv::compareHist(hist1, hist2, cv::HISTCMP_BHATTACHARYYA);
    return histDist * 1000.0;
}
//геттер для получения имени метрики
std::string TextureMetric::getName() const {
    return "texture";
}
//класс MosaicGenerator - класс для создания мозаики
//сеттер метрики по имени
bool MosaicGenerator::setMetric(const std::string& metricName) {
    if (metricName == "color") {
        metric = std::make_unique<ColorMetric>();
    }
    else if (metricName == "color_contrast") {
        metric = std::make_unique<ColorContrastMetric>();
    }
    else if (metricName == "gradient") {
        metric = std::make_unique<GradientMetric>();
    }
    else if (metricName == "texture") {
        metric = std::make_unique<TextureMetric>();
    }
    else {
        return false;
    }

    //пересчитываем параметры для всех уже загруженных тайлов
    for (auto& tile : tiles) {
        computeTileFeatures(tile, tile.image);
    }

    return true;
}

//вычисление признаков для тайла с помощью текущей метрики
void MosaicGenerator::computeTileFeatures(Tile& tile, const cv::Mat& image) const {
    if (metric) {
        metric->computeTileFeatures(tile, image);
    }
}

//загрузка тайлов из указанной папки
bool MosaicGenerator::loadTiles(const fs::path& folder, int size, bool enableRotation, int rotation) {
    //метрика по умолчанию
    if (!metric) {
        setMetric("color");
    }

    tiles.clear();
    int originalIndex = 0;
    //обход всех файлов в указанной папке
    for (const auto& entry : fs::directory_iterator(folder)) {
        if (entry.is_regular_file()) {
            cv::Mat originalTile = cv::imread(entry.path().string(), cv::IMREAD_COLOR);

            if (!originalTile.empty()) {
                //изменение размера тайла до заданного
                cv::Mat resizedTile;
                cv::resize(originalTile, resizedTile, cv::Size(size, size));
                //определение угла поворота для тайла
                std::vector<int> angles;
                if (enableRotation) {
                    angles.push_back(rotation);
                }
                else {
                    angles.push_back(0);
                }
                //поворот всех тайтлов на заданный угол
                for (int angle : angles) {
                    cv::Mat rotatedTile;
                    if (angle != 0) {
                        cv::Point2f center((float)size / 2, (float)size / 2);
                        cv::Mat rot_mat = cv::getRotationMatrix2D(center, angle, 1.0);
                        cv::warpAffine(resizedTile, rotatedTile, rot_mat, resizedTile.size(),
                            cv::INTER_LINEAR, cv::BORDER_CONSTANT, cv::Scalar(0, 0, 0));
                    }
                    else {
                        rotatedTile = resizedTile.clone();
                    }
                    //создание и настройка нового тайла
                    Tile newTile;
                    newTile.image = rotatedTile;
                    newTile.angle = angle;
                    newTile.originalIndex = originalIndex;
                    //вычисление параметров для тайла
                    computeTileFeatures(newTile, newTile.image);
                    tiles.push_back(newTile);
                }
                originalIndex++;
            }
        }
    }
    return !tiles.empty();
}

//создание мозаики без постобработки
cv::Mat MosaicGenerator::createRawMosaic(const cv::Mat& source, const Config& cfg) {
    //метрика по умолчанию
    if (!metric) setMetric("color");

    int targetWidth = source.cols;
    int targetHeight = source.rows;
    //создание пустого изображения для мозаики
    cv::Mat rawMosaic(targetHeight, targetWidth, source.type(), cv::Scalar(0, 0, 0));
    //обработка изображения по клеткам сетки
    for (int y = 0; y < targetHeight; y += cfg.gridStep) {
        for (int x = 0; x < targetWidth; x += cfg.gridStep) {
            Tile currentCell;
            //определение размеров текущей клетки
            int blockWidth = std::min(cfg.gridStep, targetWidth - x);
            int blockHeight = std::min(cfg.gridStep, targetHeight - y);

            cv::Rect region(x, y, blockWidth, blockHeight);
            cv::Mat cellImage = source(region);
            //вычисление признаков для текущей клетки
            metric->computeCellFeatures(currentCell, cellImage);

            //поиск наилучшего тайла по индексу в оригинальном векторе
            int bestIndex = -1;
            double bestDistance = std::numeric_limits<double>::max();

            for (int i = 0; i < tiles.size(); ++i) {
                if (tiles[i].usage < cfg.maxRepeats) { 
                double dist = metric->distance(currentCell, tiles[i]);
                if (dist < bestDistance) {
                    bestDistance = dist;
                    bestIndex = i;
                }
                }
            }
            //если не найден подходящий тайл, используем средний цвет клетки
            if (bestIndex == -1) {
                cv::Mat colorBlock(blockHeight, blockWidth, source.type(), cv::mean(cellImage));
                colorBlock.copyTo(rawMosaic(region));
                continue;
            }

            //увеличиваем счетчик использования
            tiles[bestIndex].usage++;
            //изменение размера тайла и копирование в мозаику
            cv::Mat finalTile;
            cv::resize(tiles[bestIndex].image, finalTile, cv::Size(blockWidth, blockHeight), 0, 0, cv::INTER_CUBIC);
            finalTile.copyTo(rawMosaic(region));
        }
    }
    return rawMosaic;
}

//создание итоговой мозаики с постобработкой
cv::Mat MosaicGenerator::createMosaic(const cv::Mat& source, const Config& cfg) {
    //проверка наличия загруженных тайлов
    if (tiles.empty()) throw std::runtime_error("No tiles loaded");
    //установка метрики сравнения
    if (!setMetric(cfg.metric)) {
        throw std::runtime_error("Invalid metric name specified: " + cfg.metric);
    }
    //сброс счетчиков использования тайтла перед генерацией
    for (auto& tile : tiles) {
        tile.usage = 0;
    }
    //создание мозаики и применение постобработки
    cv::Mat rawMosaic = createRawMosaic(source, cfg);
    return postProcessor.process(rawMosaic, source);
}

File: mosaic_processor.py
import cv2
import numpy as np
import os
from typing import List, Tuple, Optional
import math

class FeatureUtils:
    @staticmethod
    def compute_std_dev(image: np.ndarray) -> Tuple[float, ...]:
        """Вычисляет стандартное отклонение (контрастность) изображения по каналам"""
        if image.size == 0:
            return (0.0,) * image.shape[2] if len(image.shape) == 3 else (0.0,)
        
        mean, stddev = cv2.meanStdDev(image)
        
        if len(image.shape) == 3 and image.shape[2] == 3:
            return (float(stddev[0, 0]), float(stddev[1, 0]), float(stddev[2, 0]))
        return (float(stddev[0, 0]),)

    @staticmethod
    def compute_gradient_hist(image: np.ndarray) -> np.ndarray:
        """Вычисляет гистограмму направлений градиентов (HOG-подобный признак)"""
        if image.size == 0 or image.shape[0] < 3 or image.shape[1] < 3:
            return np.zeros((36, 1), dtype=np.float32)
        
        # Преобразование в оттенки серого
        if len(image.shape) == 3 and image.shape[2] == 3:
            gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        else:
            gray = image
        
        gray = gray.astype(np.float32)
        
        # Вычисление градиентов
        grad_x = cv2.Sobel(gray, cv2.CV_32F, 1, 0, ksize=3)
        grad_y = cv2.Sobel(gray, cv2.CV_32F, 0, 1, ksize=3)
        
        # Преобразование в полярные координаты
        magnitude, angle = cv2.cartToPolar(grad_x, grad_y, angleInDegrees=True)
        
        # Создание гистограммы
        hist_size = 36
        hist = np.zeros((hist_size, 1), dtype=np.float32)
        angle_step = 360.0 / hist_size
        
        for y in range(angle.shape[0]):
            for x in range(angle.shape[1]):
                angle_val = angle[y, x]
                mag_val = magnitude[y, x]
                bin_idx = int(angle_val / angle_step)
                bin_idx = max(0, min(bin_idx, hist_size - 1))
                hist[bin_idx] += mag_val
        
        # Нормализация
        cv2.normalize(hist, hist, 1.0, 0.0, cv2.NORM_L1)
        return hist.astype(np.float32)

    @staticmethod
    def get_lbp_value(gray: np.ndarray, r: int, c: int) -> int:
        """Вычисляет LBP-значение для пикселя"""
        center = gray[r, c]
        value = 0
        
        # Сравнение с 8 соседями
        neighbors = [
            (r-1, c-1), (r-1, c), (r-1, c+1),
            (r, c+1), (r+1, c+1), (r+1, c),
            (r+1, c-1), (r, c-1)
        ]
        
        for i, (nr, nc) in enumerate(neighbors):
            if gray[nr, nc] > center:
                value |= (1 << (7 - i))
        
        return value

    @staticmethod
    def compute_lbp_features(image: np.ndarray) -> np.ndarray:
        """Вычисляет LBP-признаки для описания текстуры изображения"""
        if image.size == 0 or image.shape[0] < 3 or image.shape[1] < 3:
            return np.zeros((256, 1), dtype=np.float32)
        
        # Преобразование в оттенки серого
        if len(image.shape) == 3 and image.shape[2] == 3:
            gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        else:
            gray = image
        
        # Создание LBP-карты
        lbp = np.zeros((gray.shape[0]-2, gray.shape[1]-2), dtype=np.uint8)
        
        for i in range(1, gray.shape[0]-1):
            for j in range(1, gray.shape[1]-1):
                lbp[i-1, j-1] = FeatureUtils.get_lbp_value(gray, i, j)
        
        # Вычисление гистограммы
        hist = cv2.calcHist([lbp], [0], None, [256], [0, 256])
        
        # Нормализация
        cv2.normalize(hist, hist, 1.0, 0.0, cv2.NORM_L1)
        return hist.astype(np.float32)


class Tile:
    def __init__(self):
        self.image: Optional[np.ndarray] = None
        self.original_image: Optional[np.ndarray] = None  # Добавляем хранение оригинала
        self.color: Optional[Tuple[float, ...]] = None
        self.stddev: Optional[Tuple[float, ...]] = None
        self.gradient_hist: Optional[np.ndarray] = None
        self.texture_features: Optional[np.ndarray] = None
        self.usage: int = 0
        self.angle: int = 0
        self.original_index: int = -1

class Config:
    def __init__(self):
        self.tile_size: int = 30
        self.grid_step: int = 30
        self.max_repeats: int = 1000000
        self.repeats: bool = False
        self.rotation: bool = False
        self.rotation_angle: int = 0
        self.metric: str = "color"


class IMetric:
    def compute_cell_features(self, cell: Tile, cell_image: np.ndarray):
        raise NotImplementedError
    
    def compute_tile_features(self, tile: Tile, tile_image: np.ndarray):
        raise NotImplementedError
    
    def distance(self, cell: Tile, tile: Tile) -> float:
        raise NotImplementedError
    
    def get_name(self) -> str:
        raise NotImplementedError


class ColorMetric(IMetric):
    def compute_cell_features(self, cell: Tile, cell_image: np.ndarray):
        cell.color = cv2.mean(cell_image)[:3]  # BGR mean
    
    def compute_tile_features(self, tile: Tile, tile_image: np.ndarray):
        tile.color = cv2.mean(tile_image)[:3]
    
    def distance(self, cell: Tile, tile: Tile) -> float:
        return np.linalg.norm(np.array(cell.color) - np.array(tile.color))
    
    def get_name(self) -> str:
        return "color"


class ColorContrastMetric(IMetric):
    def compute_cell_features(self, cell: Tile, cell_image: np.ndarray):
        cell.color = cv2.mean(cell_image)[:3]
        cell.stddev = FeatureUtils.compute_std_dev(cell_image)
    
    def compute_tile_features(self, tile: Tile, tile_image: np.ndarray):
        tile.color = cv2.mean(tile_image)[:3]
        tile.stddev = FeatureUtils.compute_std_dev(tile_image)
    
    def distance(self, cell: Tile, tile: Tile) -> float:
        color_dist = np.linalg.norm(np.array(cell.color) - np.array(tile.color))
        stddev_dist = np.linalg.norm(np.array(cell.stddev) - np.array(tile.stddev))
        return color_dist + 2.0 * stddev_dist
    
    def get_name(self) -> str:
        return "color_contrast"


class GradientMetric(IMetric):
    def compute_cell_features(self, cell: Tile, cell_image: np.ndarray):
        cell.gradient_hist = FeatureUtils.compute_gradient_hist(cell_image)
    
    def compute_tile_features(self, tile: Tile, tile_image: np.ndarray):
        tile.gradient_hist = FeatureUtils.compute_gradient_hist(tile_image)
    
    def distance(self, cell: Tile, tile: Tile) -> float:
        if (cell.gradient_hist is None or tile.gradient_hist is None or 
            cell.gradient_hist.size == 0 or tile.gradient_hist.size == 0):
            return float('inf')
        
        hist_dist = cv2.compareHist(cell.gradient_hist, tile.gradient_hist, cv2.HISTCMP_BHATTACHARYYA)
        return hist_dist * 1000.0
    
    def get_name(self) -> str:
        return "gradient"


class TextureMetric(IMetric):
    def compute_cell_features(self, cell: Tile, cell_image: np.ndarray):
        cell.texture_features = FeatureUtils.compute_lbp_features(cell_image)
    
    def compute_tile_features(self, tile: Tile, tile_image: np.ndarray):
        tile.texture_features = FeatureUtils.compute_lbp_features(tile_image)
    
    def distance(self, cell: Tile, tile: Tile) -> float:
        if (cell.texture_features is None or tile.texture_features is None or
            cell.texture_features.size == 0 or tile.texture_features.size == 0 or
            cell.texture_features.shape != tile.texture_features.shape):
            return float('inf')
        
        hist_dist = cv2.compareHist(cell.texture_features, tile.texture_features, cv2.HISTCMP_BHATTACHARYYA)
        return hist_dist * 1000.0
    
    def get_name(self) -> str:
        return "texture"


class MosaicGenerator:
    def __init__(self):
        self.tiles: List[Tile] = []
        self.metric: Optional[IMetric] = None
        self.post_processor = None  # Будет установлен позже
    
    def set_metric(self, metric_name: str) -> bool:
        metrics = {
            "color": ColorMetric,
            "color_contrast": ColorContrastMetric,
            "gradient": GradientMetric,
            "texture": TextureMetric
        }
        
        if metric_name not in metrics:
            return False
        
        self.metric = metrics[metric_name]()
        
        # Пересчитываем параметры для всех загруженных тайлов
        for tile in self.tiles:
            self._compute_tile_features(tile, tile.image)
        
        return True
    
    def _compute_tile_features(self, tile: Tile, image: np.ndarray):
        if self.metric and image is not None:
            self.metric.compute_tile_features(tile, image)
    
    def load_tiles(self, folder_path: str, size: int, enable_rotation: bool = False, rotation_angle: int = 0) -> bool:
        if not self.metric:
            self.set_metric("color")
        
        self.tiles.clear()
        original_index = 0
        
        try:
            for filename in os.listdir(folder_path):
                file_path = os.path.join(folder_path, filename)
                if os.path.isfile(file_path) and any(filename.lower().endswith(ext) for ext in ['.jpg', '.jpeg', '.png', '.bmp']):
                    original_tile = cv2.imread(file_path)
                    
                    if original_tile is not None:
                        # Изменение размера тайла
                        resized_tile = cv2.resize(original_tile, (size, size))
                        
                        # Определение углов поворота
                        angles = [rotation_angle] if enable_rotation else [0]
                        
                        for angle in angles:
                            if angle != 0:
                                center = (size / 2, size / 2)
                                rot_mat = cv2.getRotationMatrix2D(center, angle, 1.0)
                                rotated_tile = cv2.warpAffine(resized_tile, rot_mat, (size, size),
                                                            flags=cv2.INTER_LINEAR,
                                                            borderMode=cv2.BORDER_CONSTANT,
                                                            borderValue=(0, 0, 0))
                            else:
                                rotated_tile = resized_tile.copy()
                            
                            new_tile = Tile()
                            new_tile.image = rotated_tile
                            new_tile.angle = angle
                            new_tile.original_index = original_index
                            
                            self._compute_tile_features(new_tile, new_tile.image)
                            self.tiles.append(new_tile)
                        
                        original_index += 1
        except Exception as e:
            print(f"Error loading tiles: {e}")
            return False
        
        return len(self.tiles) > 0
    
    def load_tiles_optimized(self, folder_path: str, size: int, enable_rotation: bool = False, rotation_angle: int = 0) -> bool:
        """Оптимизированная загрузка тайлов для большого количества файлов"""
        if not self.metric:
            self.set_metric("color")
        
        self.tiles.clear()
        original_index = 0
        
        try:
            # Получаем список файлов
            image_files = [f for f in os.listdir(folder_path) 
                          if any(f.lower().endswith(ext) for ext in ['.jpg', '.jpeg', '.png', '.bmp'])]
            
            print(f"Found {len(image_files)} image files")
            
            for filename in image_files:
                file_path = os.path.join(folder_path, filename)
                
                try:
                    original_tile = cv2.imread(file_path)
                    
                    if original_tile is not None:
                        # Сохраняем оригинальное изображение
                        new_tile = Tile()
                        new_tile.original_image = original_tile.copy()
                        new_tile.image = cv2.resize(original_tile, (size, size))
                        new_tile.angle = 0  # По умолчанию без поворота
                        new_tile.original_index = original_index
                        
                        self._compute_tile_features(new_tile, new_tile.image)
                        self.tiles.append(new_tile)
                        
                        original_index += 1
                        
                        # Периодически выводим прогресс
                        if original_index % 500 == 0:
                            print(f"Processed {original_index} tiles...")
                            
                except Exception as e:
                    print(f"Error processing tile {filename}: {e}")
                    continue
                    
        except Exception as e:
            print(f"Error loading tiles: {e}")
            return False
        
        print(f"Successfully loaded {len(self.tiles)} tiles")
        return len(self.tiles) > 0
    
    def _apply_rotation(self, image: np.ndarray, angle: int, size: int) -> np.ndarray:
        """Применяет поворот к изображению"""
        if angle == 0:
            return image
        
        center = (size / 2, size / 2)
        rot_mat = cv2.getRotationMatrix2D(center, angle, 1.0)
        rotated_tile = cv2.warpAffine(image, rot_mat, (size, size),
                                    flags=cv2.INTER_LINEAR,
                                    borderMode=cv2.BORDER_CONSTANT,
                                    borderValue=(0, 0, 0))
        return rotated_tile
    
    def _create_raw_mosaic(self, source: np.ndarray, cfg: Config) -> np.ndarray:
        if not self.metric:
            self.set_metric("color")
        
        target_height, target_width = source.shape[:2]
        raw_mosaic = np.zeros((target_height, target_width, 3), dtype=np.uint8)
        
        # Определяем угол поворота
        rotation_angle = cfg.rotation_angle if cfg.rotation else 0
        
        for y in range(0, target_height, cfg.grid_step):
            for x in range(0, target_width, cfg.grid_step):
                current_cell = Tile()
                
                block_width = min(cfg.grid_step, target_width - x)
                block_height = min(cfg.grid_step, target_height - y)
                
                region = source[y:y+block_height, x:x+block_width]
                
                # Вычисление признаков для текущей клетки
                self.metric.compute_cell_features(current_cell, region)
                
                # Поиск наилучшего тайла
                best_index = -1
                best_distance = float('inf')
                
                for i, tile in enumerate(self.tiles):
                    if tile.usage < cfg.max_repeats:
                        dist = self.metric.distance(current_cell, tile)
                        if dist < best_distance:
                            best_distance = dist
                            best_index = i
                
                # Если не найден подходящий тайл, используем средний цвет
                if best_index == -1:
                    color_block = np.full((block_height, block_width, 3), cv2.mean(region)[:3], dtype=np.uint8)
                    raw_mosaic[y:y+block_height, x:x+block_width] = color_block
                    continue
                
                # Увеличиваем счетчик использования
                self.tiles[best_index].usage += 1
                
                # Подготавливаем тайл с учетом поворота
                best_tile = self.tiles[best_index]
                
                # Если требуется поворот, создаем повернутую версию
                if rotation_angle != 0 and best_tile.original_image is not None:
                    # Используем оригинальное изображение для поворота
                    resized_rotated = self._apply_rotation(
                        cv2.resize(best_tile.original_image, (cfg.tile_size, cfg.tile_size)), 
                        rotation_angle, 
                        cfg.tile_size
                    )
                    final_tile = cv2.resize(resized_rotated, (block_width, block_height), interpolation=cv2.INTER_CUBIC)
                else:
                    # Используем уже подготовленный тайл
                    final_tile = cv2.resize(best_tile.image, (block_width, block_height), interpolation=cv2.INTER_CUBIC)
                
                raw_mosaic[y:y+block_height, x:x+block_width] = final_tile
        
        return raw_mosaic
    
    def create_mosaic(self, source: np.ndarray, cfg: Config) -> np.ndarray:
        if not self.tiles:
            raise ValueError("No tiles loaded")
        
        if not self.set_metric(cfg.metric):
            raise ValueError(f"Invalid metric name specified: {cfg.metric}")
        
        # Сброс счетчиков использования
        for tile in self.tiles:
            tile.usage = 0
        
        # Создание мозаики
        raw_mosaic = self._create_raw_mosaic(source, cfg)
        
        # Применение постобработки
        if self.post_processor:
            return self.post_processor.process(raw_mosaic, source)
        
        return raw_mosaic
    
    def get_tiles_count(self) -> int:
        return len(self.tiles)
    
    def clear_tiles(self):
        self.tiles.clear()
    
    def set_post_process_config(self, config):
        if self.post_processor:
            self.post_processor.setup(config)